src/dbnode/storage/shard.go

```diff
@@ -2538,6 +2546,70 @@ func (s *dbShard) Repair(
 	return repairer.Repair(ctx, nsCtx, nsMeta, tr, s)
 }
 
+func (s *dbShard) AggregateTiles(
+	ctx context.Context,
+	reader fs.DataFileSetReader,
+	sourceNsID ident.ID,
+	sourceShard databaseShard,
+	opts AggregateTilesOptions,
+	wOpts series.WriteOptions,
+) error {
+	latestSourceVolume, err := sourceShard.latestVolume(opts.Start)
+	if err != nil {
+		return err
+	}
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   sourceNsID,
+			Shard:       sourceShard.ID(),
+			BlockStart:  opts.Start,
+			VolumeIndex: latestSourceVolume,
+		},
+		FileSetType: persist.FileSetFlushType,
+		//TODO add after https://github.com/chronosphereio/m3/pull/10 for proper streaming - OrderByIndex: true
+	}
+	if err := reader.Open(openOpts); err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
+	bytesReader := bytes.NewReader(nil)
+	dataPointIter := m3tsz.NewReaderIterator(bytesReader, true, encodingOpts)
```

Comment:
I think these readers are pooled, we can grab them out and return them.
```
iteratorPool.Init(func(r io.Reader, descr namespace.SchemaDescr) encoding.ReaderIterator {
        if cfg.Proto != nil && cfg.Proto.Enabled {
                return proto.NewIterator(r, descr, encodingOpts)
        }
        return m3tsz.NewReaderIterator(r, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
})
```

Reply:
This iterator here was used only to glue the read and write parts together. This is already replaced by Artem's stuff in https://github.com/m3db/m3/pull/2477 so I think there is no point in optimizing it.